RA-23-MVVM-Realization

DenisGubernatorov opened the pull request:
1) Во фрагмент рецепта передается не сам рецепт, а его ид 2) Перенос бизнес логики из фрагмента рецепта во viewModel 3) Удален не используемый класс FavoritesUtils.kt
4) Доработка фрагмента со списком избранных исходя из того что getFavorites теперь во viewModel и возвращает коллекцию int, а не строк

DenisGubernatorov committed:
RA-23-MVVM-Realization
1) Во фрагмент рецепта передается не сам рецепт, а его ид
2) Перенос бизнес логики из фрагмента рецепта во viewModel
3) Удален не используемый класс FavoritesUtils.kt
4) Доработка фрагмента со списком избранных исходя из того что getFavorites теперь во viewModel и возвращает коллекцию int, а не строк

aGroupoid commented on app/src/main/java/com/example/recipesapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -115,36 +100,33 @@ class RecipeFragment : Fragment() {
-                        Drawable.createFromStream(inputStream, null)
+    private fun initUI() {
+        recipeViewModel.rfLiveData.observe(viewLifecycleOwner) { state ->
+            state?.let {
```
Тогда лучше внутри этой лямбды напиши { state -> }, чтоб было понятное слово вместо it

aGroupoid commented on app/src/main/java/com/example/recipesapp/ui/recipes/recipe/RecipeViewModel.kt:
```diff
@@ -1,26 +1,75 @@
+            Context.MODE_PRIVATE
+        )
+        val idsAsStrings = sharedPrefs?.getStringSet(idsKey, emptySet()) ?: emptySet()
+        return idsAsStrings.mapNotNull { it.toIntOrNull() }.toHashSet()
```
Можно было не удалять ulits файл и использовать его здесь также. У тебя по итогу в viewModel избранного и здесь будет дублирующий код

aGroupoid commented on app/src/main/java/com/example/recipesapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -115,36 +100,33 @@ class RecipeFragment : Fragment() {
+                    recipe?.title ?: getString(R.string.get_recipes_error)
+                binding.recipesHeaderImg.setImageDrawable(
+                    try {
+                        binding.root.context.assets.open(RECIPE_IMAGE_PREFIX + recipe?.imageUrl)
```
context лучше получать через requireActivity, а не binding.root

aGroupoid commented on app/src/main/java/com/example/recipesapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -115,36 +100,33 @@ class RecipeFragment : Fragment() {
-
-            favoritesUtils.setFavorites(context, ids)
+            recipeViewModel.onFavoritesClicked(recipe?.id ?: return@setOnClickListener)
             setFavoritesButtonImage()
```
Кнопка должна только триггерить viewModel, установка фото должна происходить через обработку стейта внутри observe

aGroupoid commented on app/src/main/java/com/example/recipesapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -115,36 +100,33 @@ class RecipeFragment : Fragment() {
+            state?.let {
+                isFavorite = it.isFavorite
+                sliderPositionState = it.seekBarProgress
+                recipe = it.recipe
```
От всех этих глобальных переменных следует избавиться. Они дублируют стейт из viewModel и мы не можем гарантировать, что они всегда будут равны. Следить за двумя стейтами - лишняя морока, всё динамическое должно находится в observe, ссылаясь на данные из стейта

aGroupoid commented on app/src/main/java/com/example/recipesapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -115,36 +100,33 @@ class RecipeFragment : Fragment() {
-                Log.e("!!!!__", "image not found ${recipe?.title}", e)
-                null
+                )
+                setFavoritesButtonImage()
```
Сюда передавать значение булевое

aGroupoid commented on app/src/main/java/com/example/recipesapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -115,36 +100,33 @@ class RecipeFragment : Fragment() {
-                null
+                )
+                setFavoritesButtonImage()
+                initRecycler()
```
При каждом нажатии на кнопку, добавляется дивайдер, размер списка увеличвается. Инициализация должна происходить только 1 раз, поэтому её нужно вынести на пределы observe

DenisGubernatorov committed:
RA-23-MVVM-Realization
1) Изменена логика инициализации recycler
2) Убраны глобальные свойства во фрагменте рецепта. Все устанавливается в обсеревере
3) Вернул FavoritesUtils.kt

aGroupoid reviewed (approved):
По коду всё верно, молодец
